python/imageutils.py

```diff
@@ -471,6 +291,132 @@ def file_access(self):
+
+    read_time = 0
+
+    if remaining_mem is not None:
+        remaining_mem = remaining_mem - (split_bytes)
+
+    if remaining_mem >= 0 or remaining_mem is None:
+
+        read_s = time()     
+        data = split.get_data()
+
+        if split_shape[0] == y_size and split_shape[1] == z_size:
+            data_dict[split.header.single_vox_offset + bytes_per_voxel*(int(split_pos[-3]) + (int(split_pos[-2])) * y_size + (int(split_pos[-1]))*y_size*z_size)] = data.flatten('F')
+        else:
+            for i in range(0, split_shape[2]):
+                for j in range(0, split_shape[1]):
+                    key = split.header.single_vox_offset + bytes_per_voxel*(int(split_pos[-3]) + (int(split_pos[-2]) + j) * y_size + (int(split_pos[-1]) + i)*y_size*z_size)
+                    data_dict[key] = data[..., j, i].flatten('F')
+
+        read_time =  time() - read_s
+        print "Read time: ", read_time
+
+    else:
+        return split_name, remaining_mem, read_time
+
+    return None, remaining_mem, read_time
+
+def write_to_file(data_dict, reconstructed, bytes_per_voxel):
+
+    """Write dictionary data to file
+
+    Keyword arguments:
+
+    data_dict                           : dictionary containing contiguous data chunks as key-value pairs (key is position in reconstructed in bytes)
+    reconstructed                       : file object pointing to the to-be reconstructed image
+    bytes_per_voxel                     : number of bytes contained in a voxel in the reconstructed image
+
+    """
+
+    #defragment data before writing to file
+    defrag_time = defrag_dict(data_dict, bytes_per_voxel)
+
+    print len(data_dict)
+
+    seek_time = 0
+    write_time = 0
+
+    print "Writing data"
+    for k in sorted(data_dict.keys()):
+        seek_start = time()
+        reconstructed.seek(k, 0)
+        seek_time += time() - seek_start
+            
+        write_start = time()
+        reconstructed.write(data_dict[k].tobytes('F'))
+        write_time += time() - write_start
+        
+        #can remove from dictionary
+        del data_dict[k]
+
+    print "Seek time: ", seek_time
+    print "Write time: ", write_time
+
+    return seek_time, defrag_time, write_time
+
+def defrag_dict(data_dict, bytes_per_voxel):
+    """ Defragment dictionary containing contiguous data as key-value pairs such that key-value pairs that remain do not contain any
+        data that is contiguous to another key-value pair.
+
+        Keyword arguments:
+
+        data_dict                               : dictionary containing contiguous data chunks as key-value pairs (key is position in reconstructed in bytes)
+        bytes_per_voxel                         : number of bytes in a voxel in the reconstructed image
+
+
+    """
+
+    def_s = time()
+    previous_k = None
+    print "Dictionary size: ", len(data_dict)
+    for k in sorted(data_dict.keys()):
+        if previous_k is None:
```

Comment:
is this if statement necessary? I don't see k being equal to previous_k + ... if previous_k is None.